gms/gossiper.cc

```diff
@@ -775,6 +730,84 @@ future<gossiper::endpoint_permit> gossiper::lock_endpoint(inet_address ep) {
     });
 }
 
+
+future<> gossiper::failure_detector_thread_for_node(gms::inet_address node) {
+    auto last = gossiper::clk::now();
+    auto max_duration = std::chrono::milliseconds(_cfg.failure_detector_timeout_in_ms());
+    auto diff = std::chrono::milliseconds(0);
+    auto nr = _live_endpoints.size();
+    while (is_enabled()) {
+        bool failed = false;
+        try {
+            logger.debug("failure_detector_thread: Send echo to node {}", node);
+            co_await _messaging.send_gossip_echo(netw::msg_addr(node), max_duration);
+            logger.debug("failure_detector_thread: Send echo to node {}, status = ok", node);
+        } catch (...) {
+            failed = true;
+            logger.info("failure_detector_thread: Send echo to node {}, status = failed: {}", node, std::current_exception());
+        }
+        auto now = gossiper::clk::now();
+        diff = now - last;
+        if (!failed) {
+            last = now;
+        }
+        if (diff > max_duration) {
```

Comment:
Why are you not checking failed? And can this not cause convict() to be called twice?

Reply:
we do not care about if the verb failed or not, we care about the max time since we talked to the peer node, since the diff -> max_duration, we convict the node and exit the loop. I do not see how we can convict() twice.

Reply:
We could rely on the timeouts alone, just subtract the period from the timeout. It would make the code a bit simpler.

Reply:
Isn't the connection reused for multiple messages. If a connection failure happens, the rpc send is supposed to return immediately instead of waiting for timeout, no? In this case, we have chance to retry since we still have time budget left. If the send timeout because packet drop or other reasons, it does not matter, we think there is network issue and mark the node as down. 

Yes, we need to handle max_duration < echo_interval case. Good catch.

Reply:
> Isn't the connection reused for multiple messages. If a connection failure happens, the rpc send is supposed to return immediately instead of waiting for timeout, no? In this case, we have chance to retry since we still have time budget left. If the send timeout because packet drop or other reasons, it does not matter, we think there is network issue and mark the node as down.
> 
Ah, yes. I always forget that we are talking about two way messages here.

Reply:
Yes, I sent v7 and fixed the max_duration < echo_interval case.

Reply:
I am still not sure that calculations are exact here. Suppose first `send_gossip_echo` fails after 21 seconds. We still do not consider the peer to be down (< max_duration), we sleep for 2 seconds and then try another `send_gossip_echo` with timeout of 22 seconds. This one fails with timeout after 22 seconds and we convict the peer, but it happens after 43 seconds instead of 20.

Reply:
Since max_duration = 22 seconds and we set send_gossip_echo to timeout at max_duration, so send_gossip_echo is not supposed to return after 21 seconds, failing. If send_gossip_echo takes 21 seconds to return successfully, it is fine, it does not exceed the max_duration, we think the last echo is in time and ok.   

I see send_gossip_echo can return with failed status immediately if the peer refuses the connection, in a very unfortunate case, it can take 21 second retrying connecting, in the last second before tiemout, the connection is established but it takes 22 seconds to  timeout. In this case, the convict can be 43 seconds. I think this is acceptable for such corner case. It takes longer but it still happens in a bounded time.

Reply:
It can easily happen if a node crashed. First send will fail because of a crash and second will timeout because a connection to a dead node cannot be established.

Reply:
When a node is crashed, isn't tcp supposed to do retry until rpc times out.

This is what happens when the second node is crashed (kill the whole vm forcely). The node times out and marks the node down in 22 seconds as expected. 
```
INFO  2021-05-21 00:18:21,462 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = started
INFO  2021-05-21 00:18:21,463 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = ok
INFO  2021-05-21 00:18:23,466 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = started
INFO  2021-05-21 00:18:45,463 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = failed: seastar::rpc::timeout_error (rpc call timed out)
INFO  2021-05-21 00:18:45,463 [shard 0] gossip - failure_detector_loop: Mark node 192.168.66.120 as DOWN
INFO  2021-05-21 00:18:45,463 [shard 0] gossip - InetAddress 192.168.66.120 is now DOWN, status = NORMAL
ERROR 2021-05-21 00:19:52,034 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: No route to host

```

This is what happens when the second scylla process is terminated (kill -9 scylla). The node is marked as down in around 22 seconds too.

```
INFO  2021-05-21 00:22:43,034 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = started
INFO  2021-05-21 00:22:43,035 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = ok
ERROR 2021-05-21 00:22:44,310 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
INFO  2021-05-21 00:22:45,044 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = started
ERROR 2021-05-21 00:22:45,045 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
INFO  2021-05-21 00:22:45,045 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = failed: seastar::rpc::closed_error (connection is closed)
ERROR 2021-05-21 00:22:45,317 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
ERROR 2021-05-21 00:22:46,318 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
INFO  2021-05-21 00:22:47,048 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = started
ERROR 2021-05-21 00:22:47,049 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
INFO  2021-05-21 00:22:47,049 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = failed: seastar::rpc::closed_error (connection is closed)
ERROR 2021-05-21 00:22:47,324 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
ERROR 2021-05-21 00:22:48,333 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
INFO  2021-05-21 00:22:49,050 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = started
ERROR 2021-05-21 00:22:49,050 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
INFO  2021-05-21 00:22:49,050 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = failed: seastar::rpc::closed_error (connection is closed)
ERROR 2021-05-21 00:22:49,338 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
ERROR 2021-05-21 00:22:50,347 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
INFO  2021-05-21 00:22:51,052 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = started
ERROR 2021-05-21 00:22:51,053 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
INFO  2021-05-21 00:22:51,053 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = failed: seastar::rpc::closed_error (connection is closed)
ERROR 2021-05-21 00:22:51,349 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
ERROR 2021-05-21 00:22:52,360 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
INFO  2021-05-21 00:22:53,053 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = started
ERROR 2021-05-21 00:22:53,054 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
INFO  2021-05-21 00:22:53,054 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = failed: seastar::rpc::closed_error (connection is closed)
ERROR 2021-05-21 00:22:53,365 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
ERROR 2021-05-21 00:22:54,367 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
INFO  2021-05-21 00:22:55,057 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = started
ERROR 2021-05-21 00:22:55,058 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
INFO  2021-05-21 00:22:55,058 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = failed: seastar::rpc::closed_error (connection is closed)
ERROR 2021-05-21 00:22:55,376 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
ERROR 2021-05-21 00:22:56,384 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
INFO  2021-05-21 00:22:57,058 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = started
ERROR 2021-05-21 00:22:57,059 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
INFO  2021-05-21 00:22:57,059 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = failed: seastar::rpc::closed_error (connection is closed)
ERROR 2021-05-21 00:22:57,388 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
ERROR 2021-05-21 00:22:58,395 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
INFO  2021-05-21 00:22:59,064 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = started
ERROR 2021-05-21 00:22:59,065 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
INFO  2021-05-21 00:22:59,065 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = failed: seastar::rpc::closed_error (connection is closed)
ERROR 2021-05-21 00:22:59,405 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
ERROR 2021-05-21 00:23:00,410 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
INFO  2021-05-21 00:23:01,069 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = started
ERROR 2021-05-21 00:23:01,069 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
INFO  2021-05-21 00:23:01,070 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = failed: seastar::rpc::closed_error (connection is closed)
ERROR 2021-05-21 00:23:01,421 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
ERROR 2021-05-21 00:23:02,423 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
INFO  2021-05-21 00:23:03,072 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = started
ERROR 2021-05-21 00:23:03,072 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
INFO  2021-05-21 00:23:03,072 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = failed: seastar::rpc::closed_error (connection is closed)
ERROR 2021-05-21 00:23:03,426 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
ERROR 2021-05-21 00:23:04,427 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
INFO  2021-05-21 00:23:05,074 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = started
ERROR 2021-05-21 00:23:05,075 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
INFO  2021-05-21 00:23:05,075 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = failed: seastar::rpc::closed_error (connection is closed)
INFO  2021-05-21 00:23:05,075 [shard 0] gossip - failure_detector_loop: Mark node 192.168.66.120 as DOWN
INFO  2021-05-21 00:23:05,076 [shard 0] gossip - InetAddress 192.168.66.120 is now DOWN, status = NORMAL
ERROR 2021-05-21 00:23:05,437 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
ERROR 2021-05-21 00:23:06,437 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused

```

Reply:
Try killing a process and then immediately kill the VM.

Reply:
Here it is. It is similar, ~20s detection time. 

```
INFO  2021-05-23 23:52:07,650 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = ok                                           
INFO  2021-05-23 23:52:09,663 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = started                                      
ERROR 2021-05-23 23:52:09,664 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused                                                   
INFO  2021-05-23 23:52:09,664 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = failed: seastar::rpc::closed_error (connection is closed)
ERROR 2021-05-23 23:52:09,863 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused                                                   
ERROR 2021-05-23 23:52:10,866 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused                                                   
INFO  2021-05-23 23:52:11,669 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = started                                                                                                             
ERROR 2021-05-23 23:52:11,670 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused                                                   
INFO  2021-05-23 23:52:11,670 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = failed: seastar::rpc::closed_error (connection is closed)
ERROR 2021-05-23 23:52:11,874 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused                                                   
INFO  2021-05-23 23:52:13,671 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = started                                      
ERROR 2021-05-23 23:52:27,943 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused                                                                                                                          
INFO  2021-05-23 23:52:27,943 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = failed: seastar::rpc::closed_error (connection is closed)
ERROR 2021-05-23 23:52:27,979 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
ERROR 2021-05-23 23:52:28,989 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused             
INFO  2021-05-23 23:52:29,948 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = started
ERROR 2021-05-23 23:52:29,948 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused                                                   
INFO  2021-05-23 23:52:29,949 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = failed: seastar::rpc::closed_error (connection is closed)
INFO  2021-05-23 23:52:29,949 [shard 0] gossip - failure_detector_loop: Mark node 192.168.66.120 as DOWN
INFO  2021-05-23 23:52:29,949 [shard 0] gossip - InetAddress 192.168.66.120 is now DOWN, status = NORMAL
```

Reply:
You killed the VM after the node was already declared dead which is evident by `Connection refused` error just before the node is marked as down. Kill it before that.

Reply:
You requested: Try killing a process and then immediately kill the VM.

I did exactly that. First kill scylla process, then kill the VM.

Reply:
We are talking about timing here, so it is important _when_ you kill the VM. If you kill it after the node was already declared down then killing it achieve nothing. It is easy to see from the code that if you kill the VM before the node is declared down the connection establishment will wait for the full timeout which means that the detection time will be time between killing the Scylla process and killing the VM (which has to be less than timeout) + full timeout.

Reply:
The node was being after marked as down in about 20 seconds. I killed the VM right after I run kill -9 scylla. It can not be that the vm was killed after the node was declared as down. 

Anyway, to get a longer limited timeout, we need a very dedicated timing that requires both scylla process being killed and the machine being killed. I would not call "easily happen if a node crashed".

Reply:
You can see from the log that this is not the case though:

```
ERROR 2021-05-23 23:52:29,948 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused                                                   
INFO  2021-05-23 23:52:29,949 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = failed: seastar::rpc::closed_error (connection is closed)
INFO  2021-05-23 23:52:29,949 [shard 0] gossip - failure_detector_loop: Mark node 192.168.66.120 as DOWN
INFO  2021-05-23 23:52:29,949 [shard 0] gossip - InetAddress 192.168.66.120 is now DOWN, status = NORMAL

```

You cannot get `Connection refused ` from killed VM.

Reply:
But, I first killed the scyllal process. Can't we "Connection refused" after scylla process is killed?

Reply:
Sure. This is exactly what the log shows. You killed the process. You got `Connection refused` for 20 seconds and declared the node as done and only then you killed the VM.

Reply:
No. I did not do that. As I said, I killed the VM manually **immediately** after I press enter 'kill -9 scylla'.

Reply:
OK. I think know what is going on. I pressed the "Reset" button to kill the VM. The VM will restart. The following is what happens when I pressed "Close -> Power off".  We got a longer mark down time, but bounded as expected. 

```
INFO  2021-05-24 07:56:58,227 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = ok
INFO  2021-05-24 07:57:00,227 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = started
INFO  2021-05-24 07:57:00,228 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = ok
INFO  2021-05-24 07:57:02,235 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = started
ERROR 2021-05-24 07:57:02,236 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
INFO  2021-05-24 07:57:02,236 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = failed: seastar::rpc::closed_error (connection is closed)
ERROR 2021-05-24 07:57:02,434 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
ERROR 2021-05-24 07:57:03,446 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
INFO  2021-05-24 07:57:04,239 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = started
ERROR 2021-05-24 07:57:04,240 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
INFO  2021-05-24 07:57:04,240 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = failed: seastar::rpc::closed_error (connection is closed)
ERROR 2021-05-24 07:57:04,454 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection refused
INFO  2021-05-24 07:57:06,241 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = started
INFO  2021-05-24 07:57:28,237 [shard 0] gossip - failure_detector_loop: Send echo to node 192.168.66.120, status = failed: seastar::rpc::timeout_error (rpc call timed out)
INFO  2021-05-24 07:57:28,237 [shard 0] gossip - failure_detector_loop: Mark node 192.168.66.120 as DOWN
INFO  2021-05-24 07:57:28,238 [shard 0] gossip - InetAddress 192.168.66.120 is now DOWN, status = NORMAL
ERROR 2021-05-24 07:57:28,238 [shard 0] rpc - client 192.168.66.120:7000: fail to connect: Connection reset by peer

```